### ISSUE TITLE
Some inputs do not behave well when the run method is called repeatedly.

### ISSUE THREAD

**Issue by guyboertje** [elastic/logstash#4274]
The file input is problematic when run is called more than once, it reopens the files and file handles.  In `pipeline.rb` the method `inputworker` will call `plugin.run` if this raises an error the rescue clause will call retry - this calls the `run` method again.

Proposal: We should call a method on the input that puts it in a `re-runnable` state.

Concerns:
- We do not have a consistent API for this.  We have new, register, run, stop.  We don't have methods specifically to prepare an input for re-run.
- We could make this the responsibility of each input to detect a second call to run and do what is needed to prepare for the next run.
- We are considering a change to layered exception handling.  This may change how often a re-run might occur.

**Comment by purbon** [comment]
@guyboertje This makes complete sense to me because at the end of the day there are exception that will be out of the scope of plugin to recover. I'm not sure if we should make this a responsibility of the plugin as is really not a plugin bussiness, I see it more on the side of the pipeline.

On the other side, yeah! this will involve another big refactoring, but makes complete sense to me in order to achieve more stability.

**Comment by purbon** [comment]
@guyboertje I wonder if you can make a list containing the scope of this changes required? Is this only affecting the file input? what do you think?

**Comment by purbon** [comment]
I got another example, the twitter input (https://github.com/logstash-plugins/logstash-input-twitter/issues/7#issuecomment-160921799), still under investigation but it do really looks like this is the expected problem.

**Comment by colinsurprenant** [comment]
wouldn't be simpler to have the `close` method make sure the input plugin is left in a re-runnable state?

**Comment by purbon** [comment]
@colinsurprenant I think we should define a proper semantic, sometimes is not easy to close having a re runnable state (closing sockets for example). I think most of the times register is doing just too much.

**Comment by colinsurprenant** [comment]
Agree, I think it is more a question of defining the proper semantic around the current API with `register`, `run` and `close`. I don't think we necessarily need to introduce a new API?

**Comment by purbon** [comment]
My feeling now is we should break register, at less until we "change" the way we handle the input worker (calling close before rerun, but not register again).
